bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/ReviewDataSourceServlet.java

```diff
@@ -0,0 +1,126 @@
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.forms.core.components.internal.servlets;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.servlet.Servlet;
+
+import org.apache.jackrabbit.JcrConstants;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.SlingHttpServletResponse;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.SyntheticResource;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.jetbrains.annotations.NotNull;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import com.adobe.cq.export.json.ComponentExporter;
+import com.adobe.cq.forms.core.components.internal.form.FormConstants;
+import com.adobe.cq.forms.core.components.models.form.Base;
+import com.adobe.cq.forms.core.components.models.form.Container;
+import com.adobe.cq.forms.core.components.models.form.FormComponent;
+import com.adobe.cq.forms.core.components.models.form.FormContainer;
+import com.adobe.cq.forms.core.components.models.form.Label;
+import com.adobe.cq.forms.core.components.util.ComponentUtils;
+import com.adobe.granite.ui.components.ExpressionResolver;
+import com.adobe.granite.ui.components.ds.DataSource;
+import com.adobe.granite.ui.components.ds.SimpleDataSource;
+import com.adobe.granite.ui.components.ds.ValueMapResource;
+
+@Component(
+    service = { Servlet.class },
+    property = {
+        "sling.servlet.resourceTypes=" + FormConstants.RT_FD_FORM_REVIEW_DATASOURCE_V1,
+        "sling.servlet.methods=GET",
+        "sling.servlet.extensions=html"
+    })
+public class ReviewDataSourceServlet extends AbstractDataSourceServlet {
+
+    /**
+     * Defines the form meta data type. Possible values: {@code submitAction},
+     * {@code prefillServiceProvider}
+     *
+     * @todo: Add other metadata types here like fragment, actions etc
+     */
+    @Reference
+    private transient ExpressionResolver expressionResolver;
+
+    @NotNull
+    @Override
+    protected ExpressionResolver getExpressionResolver() {
+        return expressionResolver;
+    }
+
+    @Override
+    protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) {
+
+        ResourceResolver resourceResolver = request.getResourceResolver();
+        String componentInstancePath = request.getRequestPathInfo().getSuffix();
+        List<Resource> resources = new ArrayList<>();
+        if (resourceResolver != null) {
```

Comment:
We should change the implementation by using FormContainer#visit, like we do during form submission, that way could be more readable.

In the current implementation, we were walking twice which is not required.